src/dune_rpc_impl/server.ml

```diff
@@ -119,34 +119,97 @@ end)
 
 module Session_set = Session_comparable.Set
 
-module Subscribers = struct
-  type t =
-    { build_progress : Session_set.t
-    ; diagnostics : Session_set.t
-    }
+module Subscribers : sig
+  type t
+
+  type _ what =
+    | Diagnostic : Diagnostic.Event.t list what
+    | Progress : Progress.t what
+
+  val empty : t
+
+  val remove : t -> 'a what -> 'a Subscription.t -> t
 
-  let empty =
-    { build_progress = Session_set.empty; diagnostics = Session_set.empty }
+  (** Remove all subscriptions of a sesssion *)
+  val remove_session : t -> Client.t Session.t -> t
 
-  let get t (which : Subscribe.t) =
-    match which with
-    | Build_progress -> t.build_progress
-    | Diagnostics -> t.diagnostics
+  val add : t -> 'a what -> 'a Subscription.t -> t
+
+  val notify : t -> 'a what -> 'a -> unit Fiber.t
+end = struct
+  module S = struct
+    let compare = Subscription.compare
+
+    let to_dyn x = Subscription.to_dyn x
+  end
 
-  let modify t (which : Subscribe.t) ~f =
-    match which with
-    | Build_progress -> { t with build_progress = f t.build_progress }
-    | Diagnostics -> { t with diagnostics = f t.diagnostics }
+  module Progress_subs_comparable = Comparable.Make (struct
+    type t = Progress.t Subscription.t
 
-  let add t which session =
-    modify t which ~f:(fun set -> Session_set.add set session)
+    include S
+  end)
 
-  let remove t which session =
-    modify t which ~f:(fun set -> Session_set.remove set session)
+  module Progress_subs = Progress_subs_comparable.Set
 
-  let notify t which ~f =
-    let set = get t which in
-    Fiber.parallel_iter_set (module Session_set) set ~f
+  module Diagnostic_subs_comparable = Comparable.Make (struct
+    type t = Diagnostic.Event.t list Subscription.t
+
+    include S
+  end)
+
+  module Diagnostic_subs = Diagnostic_subs_comparable.Set
+  module Session_map = Session_comparable.Map
+
+  module Sub_comparable = Comparable.Make (struct
+    type t = Subscription.packed
+
+    let compare x y = Subscription.compare_packed x y
+
+    let to_dyn (Subscription.E x) = Subscription.to_dyn x
+  end)
+
+  module Sub_map = Sub_comparable.Map
+
+  type t =
+    { build_progress : Progress.t Subscription.t Sub_map.t
+    ; diagnostic : Diagnostic.Event.t list Subscription.t Sub_map.t
+    }
+
+  type _ what =
+    | Diagnostic : Diagnostic.Event.t list what
+    | Progress : Progress.t what
+
+  let empty = { build_progress = Sub_map.empty; diagnostic = Sub_map.empty }
+
+  let add (type a) t (what : a what) (sub : a Subscription.t) =
+    let key = Subscription.E sub in
+    match what with
```

Comment:
Maybe you can avoid most of these pattern matches by implementing something like this?

```
val with_sub_map : 'a what -> t -> ('a Subscription.t Sub_map.t -> ('a Subscription.t Sub_map.t * 'b)) -> (t * 'b)
```

At the very it will decouple the "homemade Univ_map" logic from the actual subscription management logic.